packages/scale-codec/src/codecs/Map.ts

```diff
@@ -0,0 +1,24 @@
+import { Codec, Decoder, Encoder } from "../types"
+import { createCodec, enhanceDecoder, enhanceEncoder } from "../utils"
+import { Tuple } from "./Tuple"
+import { Vector } from "./Vector"
+
+const MapCodecEnc = <K, V>(key: Encoder<K>, value: Encoder<V>) =>
+  enhanceEncoder(Vector.enc(Tuple.enc(key, value)), (input: Map<K, V>) =>
+    Array.from(input.entries()),
+  )
+
+const MapCodecDec = <K, V>(key: Decoder<K>, value: Decoder<V>) =>
+  enhanceDecoder(
+    Vector.dec(Tuple.dec(key, value)),
+    (entries) => new Map(entries),
+  )
+
+export const MapCodec = <K, V>(
+  key: Codec<K>,
+  value: Codec<V>,
+): Codec<Map<K, V>> =>
+  createCodec(MapCodecEnc(key[0], value[0]), MapCodecDec(key[1], value[1]))
+
+MapCodec.enc = MapCodecEnc
```

Comment:
Sorry, I don't quite understand this comment... Could you please elaborate a little bit more? :pray:

Reply:
The encode and decode fns are attached to `MapCodec`. Instead of loaded on demand. Is this necessary?

Reply:
Right... So, yep, I'm not completely sure about this TBH, but I think that I kinda like it this way. This is how I got here:
- Initially the Codecs were just a tuple of their `[encoder, decoder]` that worked "nicely" for destructuring them. However, if you just wanted to build another encoder (or another decoder) rather than the whole codec, then it looked a bit weird... Then I thought, what if the type Codec looked like this?
```ts
type Codec<T> = [Encoder<T>, Decoder<T>] & { enc: Encoder<T>, dec: Decoder<T>}
```
that way we would have the best of both worlds, as we could do things like `string.enc('Some text')` rather than having to do `string[0]('Some text')` or 
```ts
const [stringEncoder] = string
stringEncoder('Some text')
```

And then I realized that if Codecs are going to have those 2 properties, if higher-order codecs also had them, then it would be a lot easier to compose just encoders or just decoders, for instance:

```ts
const myCustomEncoder = Struct.enc({
  foo: string.enc,
  bar: Vector.enc(boolean.enc),
})
```

Which, in reality is almost the same as doing:
```ts
const myCustomEncoder = Struct({
  foo: string,
  bar: Vector(boolean),
}).enc
```

However, in some cases it could make a lot of sense to use the `.enc` version, like if you only have the encoder without the decoder, or the other way around. For instance, imagine that I have an encoder that I've imported from somewhere else and I don't have (and don't need the decoder), then I could do:

```ts
const myCustomEncoder = Struct.enc({
  foo: theExternalEncoder,
  bar: Vector.enc(boolean.enc),
})
```

which IMO is a lot less hacky than having to do:

```ts
const myCustomEncoder = Struc({
  foo: createCodec(theExternalEncoder, noop as any),
  bar: Vector(boolean)
}).enc
```

Now, is this really necessary? Could I have Exported `Vec`, `VecEncoder` and `VecDecoder` rather than putting it all together?

Yes, of course! In fact, perhaps that would be a better way to go... but I liked the idea of having these common props (`enc` & `dec`) between the Codecs and the Higher-Order Codecs.

Does this make sense? Do you think that it would be better to get rid of the `enc`/`dec` props?

Reply:
It does make sense, yes! & no, I don't think it would make a huge difference.

One last thought to throw out there... perhaps this could be inverted. Aka. users explicitly import `Encoder` or `Decoder` and supply the struct instance.

```ts
const myStruct = t.Struct({
  a: t.str,
});

const encodeMyStruct = Encoder(myStruct);
```